### PR TITLE
Apply captured outer type parameters

### LINKED ISSUE #5400: Generic types do not propagate into inherited class expressions
When using `extends` in a class expression, the generic types do not propagate correctly to the super class of the expression.   The below example should compile (I think!), but fails with `Type 'C' is not assignable to type 'T'`.

```
class A<T> {
	genericVar: T
}

function B<T>() {
	return class extends A<T> {
	}
}

class C extends B<C>() {
	name: string;
}

var c = new C();
c.genericVar = c;
```

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #5400. 

Classes with captured type parameters cannot use `getTypeFromClassOrInterfaceReference`'s fast path of applying type arguments directly to the type reference. Currently, `number` is not applied to `<T>` in the example below.

```ts
class A<T> { }
function B<T>() {
  return class extends A<T> { }
}
var b = new B<number>();
```

This PR fixes that by instead calling the slow path: `getReturnTypeOfSignature(getInstantiatedConstructorsForTypeArguments(...`

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2913,6 +2915,21 @@ namespace ts {
             }
         }
 
+        function baseTypeHasUnappliedOuterTypeParameters(type: Type): boolean {
+            const originalBaseType = <InterfaceType>type;
+            const originalTypeReference = <TypeReference>type;
+            if (originalBaseType.outerTypeParameters) {
+                // an unapplied type type parameter is one
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: type type

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2913,6 +2915,21 @@ namespace ts {
             }
         }
 
+        function baseTypeHasUnappliedOuterTypeParameters(type: Type): boolean {
+            const originalBaseType = <InterfaceType>type;
+            const originalTypeReference = <TypeReference>type;
+            if (originalBaseType.outerTypeParameters) {
+                // an unapplied type type parameter is one 
+                // whose argument symbol is still the same as the parameter symbol
+                for (let i = 0; i < originalBaseType.outerTypeParameters.length; i++) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: cache the length

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -2913,6 +2915,21 @@ namespace ts {
             }
         }
 
+        function baseTypeHasUnappliedOuterTypeParameters(type: Type): boolean {
+            const originalBaseType = <InterfaceType>type;
+            const originalTypeReference = <TypeReference>type;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: You only use these for the type parameters and type arguments. Might as well just declare variables for those instead.

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/checker.ts]: type type type!

Fixed.

File: src/compiler/checker.ts (hunk 2 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts (hunk 3 above)
Comment (sandersn) [inline src/compiler/checker.ts]: Good idea. This bothered me, and that's a nice improvement.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -2913,6 +2915,22 @@ namespace ts {
...
         }
 
+        function baseTypeHasUnappliedOuterTypeParameters(type: Type): boolean {
+            const outerTypeParameters = (<InterfaceType>type).outerTypeParameters;
+            const typeArguments = (<TypeReference>type).typeArguments;
+            if (outerTypeParameters) {
+                // an unapplied type parameter is one 
+                // whose argument symbol is still the same as the parameter symbol
+                const numParameters = outerTypeParameters.length;
+                for (let i = 0; i < numParameters; i++) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: If it's the case that some type parameters might be applied and some might *not* be applied, are there any that are more likely to be applied than others?

In other words, do we partially apply them towards the beginning? If so, then iterating from the back would bail out more quickly if any type parameter is not applied.

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Actually, if it's the case that type parameters are always applied from earlier indices onward, you could just check at the last index.

Comment (sandersn) [inline src/compiler/checker.ts]: Good idea. I'm pretty sure that's the case, so I'll try it.

Comment (sandersn) [inline src/compiler/checker.ts]: Yes, looks like that works.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -2913,6 +2915,18 @@ namespace ts {
...
         }
 
+        function areAllOuterTypeParametersApplied(type: Type): boolean {
+            const outerTypeParameters = (<InterfaceType>type).outerTypeParameters;
+            if (outerTypeParameters) {
+                // an unapplied type parameter is one 
+                // whose argument symbol is still the same as the parameter symbol
+                const last = outerTypeParameters.length - 1;
+                const typeArguments = (<TypeReference>type).typeArguments;
+                return outerTypeParameters[last].symbol !== typeArguments[last].symbol;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Can you leave a comment about why we can get away with just checking the last type parameter? It's not necessarily clear to someone stumbling upon this code for the first time.

Comment (sandersn) [inline src/compiler/checker.ts]: done

**Comment by DanielRosenwasser** [comment]
:+1:

**Comment by russpowers** [comment]
This is great, thanks for fixing it so quickly! :+1: :+1:

**Pull request merged**